Implement `propagate_error` argument

kylesayrs opened the pull request:
## Changes ##
* Add `propagate_error` argument to sequential pipeline.
  * When True, the sequential pipeline propagates quantized outputs (outputs after module optimization/GPTQ apply)
  * When False, the sequential pipeline propagates unquantized outputs (full precision)
  * [GPTAQ](https://arxiv.org/pdf/2504.02692) authors report slightly better accuracy recovery when `propagate_error=False`. This aligns with the preliminary testing I've done myself, but more work will be required to make this a default argument.

kylesayrs committed:
implement propagate_error

Signed-off-by: Kyle Sayers <kylesayrs@gmail.com>

kylesayrs committed:
optimization

Signed-off-by: Kyle Sayers <kylesayrs@gmail.com>

kylesayrs committed:
fix typo

Signed-off-by: Kyle Sayers <kylesayrs@gmail.com>

brian-dellabetta reviewed (approved):
nice! nit -- `batch_idx` is a bit more readable than `b_idx`

kylesayrs commented:
@brian-dellabetta I agree, but I think `b_idx` is more readable than putting the loop on two separate lines 🙁

dsikka commented:
@mergifyio refresh

dsikka committed:
Merge branch 'main' into kylesayrs/propagate-error

HDCharles reviewed (approved):
see comment but assuming there's a good reason, going to accept

dsikka committed:
Merge branch 'main' into kylesayrs/propagate-error

HDCharles committed:
Merge branch 'main' into kylesayrs/propagate-error

HDCharles committed:
Merge branch 'main' into kylesayrs/propagate-error

kylesayrs commented:
@coderabbitai resolve merge conflict

coderabbitai[bot] committed:
Resolve merge conflicts in 1 file(s)

Resolved conflicts in:
- src/llmcompressor/pipelines/sequential/pipeline.py (content)

Co-authored-by: CodeRabbit <noreply@coderabbit.ai>

kylesayrs committed:
apply style

Signed-off-by: Kyle Sayers <kylesayrs@gmail.com>

kylesayrs committed:
Merge branch 'main' into kylesayrs/propagate-error

kylesayrs commented on src/llmcompressor/pipelines/sequential/pipeline.py:
```diff
@@ -104,22 +104,27 @@ def __call__(
-                    LifecycleCallbacks.sequential_epoch_end(subgraph)
+                        if not dataset_args.propagate_error:
+                            activations.update(b_idx, outputs)
+                            activations.delete(b_idx, subgraph.consumed_names)
```
We cannot move it to the bottom, since by that time the intermediates from the forward pass have been garbage collected.
